### PR TITLE
Feature/validate with missing props

### PR DESCRIPTION
**Pull request by gdimauro**
Hi, using AutoMapper I've appreciate very much a feature which on changing of entities or other mapped models the general validation method invoked during tests shows a complete list of missing fields from all unbalanced mappings. The validation method I changed in this pull adds an optional output array of string which in case of incomplete mappings shows you all missing fields in rules. Added also all necessary tests for covering feature.

### CODE REVIEW

**Comment by bchavez** [comment]
Hi @gdimauro ,

Thank you for your contribution and your valuable feedback. I very much appreciate it. I think the feature you've described can be useful in testing situations. However, I have some concerns I'd like to talk to you about.

The approach I would have suggested first is to create an issue for your feature before sending a pull request. This would at least give us (and others) a chance to have a discussion about your proposed features, API design, and changes before you invest time into something that might not get merged. Everytime a PR is closed without merging, a kitten is lost. :crying_cat_face: 

So, first off, I'd like to know how is your proposed change different from using `StrictMode` and `rule.Ingore()`:
```
var faker = new Faker<Order>()
    .StrictMode(true)
    .Ignore(o => o.Item)
    .RuleFor(o => o.OrderId, f => 3343)
    .RuleFor(o => o.Quantity, f => f.Random.Number(3));
```
Why, specifically, does this not fit your requirements? `Validation` with `.StrictMode(true)` and `.Ignore(...)` should still provide you with an assurance that your rules and domain changes are still balanced.

------

Addtionally, as far as I can tell, your suggestion is in this commit:
https://github.com/bchavez/Bogus/pull/37/commits/31f8914d4b57687dfe502537c766de3d9118c0be#diff-79ef2dbf184708cb654078f6c6f48747R60

```
string[] missingPropsOrFields;
var result = testOrders.Validate(out missingPropsOrFields);
```

Is this the ***exact*** same method parameter that **AutoMapper** uses? Can you point me to the API soure-code or docs where **AutoMapper** describes this feature?

The reason I'm asking is, I don't like having `out` parameters much because they are a pain to deal with. There's a lot of setup drama and clunkyness with `out` prams. I'd much rather have something like: `var r = testOrders.ValidateResult()` that gives back an object with `result.IsValid` and something like `result.MissingRules` instead of having to prop up a `string[] missingPropsOrFields` variable first... So, if you could point me to how **AutoMapper** resolves this issue that'd be great; I'd like to see how they do it first. Also, this might be a moot point with recent developments of C# 7 and multiple return values.

----

Third, and lastly, your pull request is way too large my friend. You only need to make changes to **Bogus**' source code. It is ***not necessarily*** upgrade project dependencies, runtimes, tooling, and unit testing platforms. I apprecate the helpful effort but it is counterproductive here. I cannot merge PRs that is this large because you are changing way too much that is not related to your specific API feature.

This leads me to the reason why, ***I think*** your initial build failed. I think it is precicely because your initial commit started changing and upgrding, tooling, dependencies, and references in project files. Honestly, all you need is ***just this*** https://github.com/bchavez/Bogus/pull/37/commits/31f8914d4b57687dfe502537c766de3d9118c0be#diff-79ef2dbf184708cb654078f6c6f48747R60

-----

I look forward hearing from you.

Thanks,
Brian

:crown: :gem: [***"I know everything that shine ain't always gold..."***](https://www.youtube.com/watch?v=7xzU9Qqdqww)

**Comment by gdimauro** [comment]
Hi Brian, thanks for the prompt response and compliments for this useful library. Automapper has a different API for testing complete coverage of mappings because of a different approach it uses for declaring only during startup of the entire application all the mapping configuration at once. For this reason for testing if some changes has been made to entities related to some kind of mapping rules you can run a single method which shows you all the eventually changes causing lacking mappings, raising an exception in a test method like you can see in the following code:

    /// <summary>
    /// Check if all mapping configuration are consistent and valid
    /// </summary>
    [Fact(DisplayName = "Complete/all mapping configuration test")]
    public void TestAllMappingConfigurations()
    {
      Mapper.AssertConfigurationIsValid();
    }

In your case you have to check every single "Faker" separately because actually you don't have a somewhere a complete list of available fixtures. But this can be a future enhancement we can talk about implementing something similar to AutoMapper in a declerative way at startup which I find very useful. So basically, the idea was to give the developer a chance to identify immediately the complete list of all uncovered fields for fixtures (because of changes to model entities etc) with strictmode explicitly or implicitly (through your global settings) set to true. Of course in that case you can use .Ignore method to excplictly ignore new fields and close the case.

Here a link to some info about automapper configuration test and an interesting explanation about this functionality from the "field":

[https://github.com/AutoMapper/AutoMapper/wiki/Configuration-validation](https://github.com/AutoMapper/AutoMapper/wiki/Configuration-validation)

http://stackoverflow.com/questions/12037986/is-automapper-assertconfigurationisvalid-enough-to-ensure-good-mapping

**Lost kittens**
OK. First time for me here. Lesson learned. Fantastic protocol: loving it more and more.

**About out param:**
OK. I agree with you. It's not clean enough because we can avoid I but, I've left previuous API intact without this new parameter/feature for backward compatibility. Obviuously I don't want to change the return value from boolean to another return value because of BW compatibilty. On the other side I like your proposal and I'll implement it ASAP and return to you. At this point I can also think to implement a method very similar to AutoMapper naming it AssertConfigurationIsValid or similar raising an exception with the list of missing rules. Internally it will use your original Validate method and raise an exception with the complete list of missing rules in order to see the reason of the failure in the test results immediately, giving the dev the opportunity to understand and fix the lack.

**Third and lastly**
Agree!

Finally, I'll return to you as soon I address all your helpful proposals. Thanks in advance, Giuseppe

**Comment by gdimauro** [comment]
BTW, this is a typical exception AutoMapper raises during tests when lacking configurations are found:

AutoMapper.AutoMapperConfigurationException occurred
  HResult=-2146233088
  Message=
Unmapped members were found. Review the types and members below.
Add a custom mapping expression, ignore, add a custom resolver, or modify the source/destination type
For no matching constructor, add a no-arg ctor, add optional arguments, or map all of the constructor parameters
============================================================
Operatore -> Operatore (Destination member list)
**omissis**.Operatore -> **omissis**.Model.Entities.Operatore (Destination member list)

Unmapped properties:
Memo
ChangeTracker
==============================================================================================
CategoriaImmobile -> CategoriaImmobile (Destination member list)
**omissis**.Fascicoli.Oggetto.CategoriaImmobile -> **omissis**.Model.Entities.CategoriaImmobile (Destination member list)

Unmapped properties:
DatiImmobile
==============================================================================================
CategoriaImmobile -> CategoriaImmobile (Destination member list)
**omissis**.Model.Entities.CategoriaImmobile -> **omissis**.Fascicoli.Oggetto.CategoriaImmobile (Destination member list)

Unmapped properties:
DatiImmobili
ChangeTracker
========================================================================================
NaturaImmobile -> NaturaImmobile (Destination member list)
**omissis**.Fascicoli.Oggetto.NaturaImmobile -> **omissis**.Model.Entities.NaturaImmobile (Destination member list)

Unmapped properties:
DatiImmobile
========================================================================================
NaturaImmobile -> NaturaImmobile (Destination member list)
**omissis**.Model.Entities.NaturaImmobile -> **omissis**.Fascicoli.Oggetto.NaturaImmobile (Destination member list)

Unmapped properties:
DatiImmobili
ChangeTracker
==================================================================================
TipoOggetto -> TipoOggetto (Destination member list)
**omissis**.Model.Entities.TipoOggetto -> **omissis**.Fascicoli.Oggetto.TipoOggetto (Destination member list)

Unmapped properties:
ChangeTracker
========================================================================================
RichiestaContatto -> Fascicolo (Destination member list)
**omissis**.Model.Entities.RichiestaContatto -> **omissis**.Fascicoli.Fascicolo.Fascicolo (Destination member list)

Unmapped properties:
Codice
DataConferimentoIncarico
Notaio
Associazione
Template
Note
StatoAvanzamento
RichiestaId
==================================================================================
Richiesta -> RichiestaContatto (Destination member list)
**omissis**.Crm.Richiesta.Richiesta -> **omissis**.Model.Entities.RichiestaContatto (Destination member list)

Unmapped properties:
IsDeleted
========================================================================
Persona -> Persona (Destination member list)
**omissis**.Model.Entities.Persona -> **omissis**.Clienti.Persona.Persona (Destination member list)

Unmapped properties:
Tipo
DatiAmministrativi
=================================================================================
PersonaGiuridica -> Persona (Destination member list)
**omissis**.Clienti.Persona.PersonaGiuridica -> **omissis**.Model.Entities.Persona (Destination member list)

Unmapped properties:
Nome
Cognome
DataNascita
FrazioneNascita
DenominazioneComuneNascita
StatoNascita
StatoNascitaId
ComuneNascita
ComuneNascitaId
ProvinciaNascita
ProvinciaNascitaSigla
Cittadinanza
TipoStatoCivile
Titolo
CheckedDittaIndividuale
RagioneSocialeDittaIndividuale
SediDittaIndividuale
TipoRegime
==========================================================================
Attivita -> Attivita (Destination member list)
**omissis**.Agenda.Attivita.Attivita -> **omissis**.Model.Entities.Attivita (Destination member list)

Unmapped properties:
Properties

  Source=AutoMapper
  StackTrace:
       at **omissis**.Mapping.Tests.NotaroMappingTests.TestMappingConfiguration() in C:\prj\**omissis**\server\**omissis**.Containers\GEN\WKI\Notaro\Notaro+Mapping+Tests.cs:line 47
  InnerException:

**Comment by bchavez** [comment]
Hi @gdimauro ,

Thank you very much for your response. I now have a better understanding of your situation. I absolutely agree, something like `AssertConfigurationIsValid` will be very useful on the `Faker` objects in testing situations. I'll support you in getting your changes into **Bogus**.

Echoing your thoughts, I think we should **not** expose `string[]` part of our public API that describes the list of missing properties / fields. Having an assertion like `AssertConfigurationIsValid` throw a detailed exception message similar to **AutoMapper**, I think, is best in terms of API aesthetics.

If you can get your feature implemented without breaking backwards compatibility, that would be fantastic, but I'm also not afraid to break API compatibility if it is absolutely necessary, in the name of better API semantics. We just need to document the breaking change in our `HISTORY.md` and increment the major version number then call it a day.

I think I see your feature doing something like:

```
internal ValidationResult ValidateInternal(){
    var result = 
}
//remains as-is, backwards compatability, no breaking changes
public bool Validate(){
    var result = ValidateInternal();
    return result.IsValid;
}
public void AssertConfigurationIsValid(){
    var result = ValidateInternal();
    if(!result.IsValid){
       throw ExceptionFormatter.ValidationException(result);
    }
}
```
Or something like that. Feel free to add any support `ExceptionFormatter` classes as you see fit. I'll leave the details up to you. Also, I think the `Generate` method checks `StrictMode` too, so we may have to put the `ExceptionFormatter` there too to make sure we are consistent.

I look forward to seeing your contribution. Thanks again for raising your issue with **Bogus** and thank you for your contribution. You can also add your name to the `README.md` part of your PR for some street cred. :sunglasses: 

Good night. :night_with_stars:.
Brian

:fallen_leaf: :leaves: [***"I get a sense of weightlessness…"***](https://www.youtube.com/watch?v=lPdNTBwRR-4)

**Comment by gdimauro** [comment]
Hi Brian, just pushed.

Cheers,
Giuseppe

**Comment by bchavez** [comment]
Hi @gdimauro ,

Thanks for pushing your changes; however, I still see 14 files changed: +2,157/−69 at the top.

The `Newtonsoft` dependencies are still downgraded to v6 from v9 of what they originally were. The next step would be to back out and revert all the previous commits that made the changes to the dependencies and tooling that affected the project files.

I don't have the ability to revert those changes on my side since you own this PR.

Hope that helps!
Brian

:boom: :fire: [***"Set it ablaze like a candle wick... Light it up, light it up..."***](https://www.youtube.com/watch?v=qDcFryDXQ7U)

**Comment by gdimauro** [comment]
Brian, thank you so much for your support and sorry for the little delay. In brief: I'm using Visual Studio 2015 and in order to execute tests and build all the stuff with VS2015 I needed to do all the modifications and upgrades you've seen. In the very next days I'll start from scratch again adding only the necessary changes for integrating my little changes. After this activity I'll return to you for in order to close this interesting PR. Cheers, Giuseppe.

**Comment by bchavez** [comment]
Hi @gdimauro ,

No problem. If it helps, you can also use the **Visual Studio 2015 Command Prompt**, in a fresh checkout directory and execute the following build commands:
* `build.cmd restore` to restore packages as they are restored on the CI server (should be done first before executing other build commands).
* `build.cmd msb` to run MSBuild and run the compiler
* `build.cmd test` to run unit tests.

:chocolate_bar: :cookie: :lollipop: [***Ronald Jenkees - Stay Crunchy***](https://www.youtube.com/watch?v=GeIAXlwVlZc)

**Comment by gdimauro** [comment]
This is exactly what I thought to do. Just for the sake, I beforehand you that I'm also working on an extension for generating fake data directly from attributes which has a major priority for me in this phase of the project I'm using your useful stuff. Just a little preview/example without rulesets which will be covered ASAP:

 [Table("TestEntity")]
  [Fixtures(strict: true)]
  public partial class TestEntityDetail : TenantBaseEntity
  {
    [DataMember]
    [MaxLength(50, ErrorMessage = "")]
    [DataType(DataType.Text, ErrorMessage = "")]
    [Fixture("@=Name.FirstName(Bogus.DataSets.Name.Gender.Male)")]
    public virtual string Name { get; set; }

    [DataMember]
    [MaxLength(50, ErrorMessage = "")]
    [DataType(DataType.Text, ErrorMessage = "")]
    [Fixture(null)]
    public virtual string NullField { get; set; }

    [DataMember]
    [Fixture(FixtureLastname)]
    public virtual string Lastname2 { get; set; }

    [DataMember]
    [Fixture(FixtureFemaleLastname)]
    public virtual string MaleLastname { get; set; }

    [DataMember]
    [Fixture(FixtureFemaleLastname)] // FixtureFemaleLastname is a constant string equals to: "@=Name.LastName(Bogus.DataSets.Name.Gender.Female)"
    public virtual string FemaleLastname { get; set; }

    [DataMember]
    [Fixture("=Guid.Parse(\"{8F55D58C-D14E-4E0A-B777-C2B427F31634}\")")]
    public virtual Guid Guid1 { get; set; }

    ///<summary>
    /// Cognome di una persona fisica
    ///</summary>
    [DataMember]
    [MaxLength(50, ErrorMessage = "")]
    [DataType(DataType.Text, ErrorMessage = "")]
    [Fixture(@"@=Name.LastName(Bogus.DataSets.Name.Gender.Male)")]
    public virtual string Lastname { get; set; }
  }

**Comment by gdimauro** [comment]
Brian, hope this is OK. Cheers, Giusppe

File: Source/Bogus/Faker[T].cs [hunk 1]
```diff
@@ -246,10 +247,10 @@ private void PopulateInternal(T instance, string[] ruleSets)
                     }
                 }
 
-                foreach( var ruleSet in ruleSets )
+                foreach (var ruleSet in ruleSets)
```
Comment (bchavez) [inline Source/Bogus/Faker[T].cs]: Hey @gdimauro , for future reference, all these space formatting edits add a lot of noise to the PR.

Comment (bchavez) [inline Source/Bogus/Faker[T].cs]: Hey @gdimauro , no worries, I'm going to rebase and squash the commit history in this PR so we have just 1 commit with only the necessary changes. Thanks again for getting your PR this far; so far it looks wonderful and very much easier for me to digest.

I'm just making a few minor changes. I'll CC you the PR once I'm done reviewing your changes.

**Comment by bchavez** [comment]
Superseded by PR #39 

Thanks @gdimauro , your changes are in `master`. =)

:art: :dash: [**Marshmello - CoLoUR**](https://www.youtube.com/watch?v=6UhNJC8OyVA)

**Comment by bchavez** [comment]
@gdimauro , your changes are now live on Nuget `Bogus v7.1.7`. Thanks again for your contribution. 

https://www.nuget.org/packages/Bogus/

Brian

**Comment by gdimauro** [comment]
Brian, thank you so much!
If you plan a trip in the south of Italy please let me know :)
Giuseppe




----Messaggio originale----

Da: "Brian Chavez" <notifications@github.com>

Data: 11/10/2016 23.20

A: "bchavez/Bogus"<Bogus@noreply.github.com>

Cc: "Giuseppe Dimauro"<gdimauro@libero.it>, "Mention"<mention@noreply.github.com>

Ogg: Re: [bchavez/Bogus] Feature/validate with missing props (#37)




@gdimauro , your changes are now live on Nuget Bogus v7.1.7. Thanks again for your contribution. 


https://www.nuget.org/packages/Bogus/


Brian

**Pull request closed without merging**